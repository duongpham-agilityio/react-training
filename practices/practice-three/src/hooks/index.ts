export * from './useProduct';
export * from './usePagination';
export * from './useDebounce';
export * from './useSearch';
