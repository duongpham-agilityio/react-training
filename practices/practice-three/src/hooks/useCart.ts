import { useCallback } from 'react';

// Constants
import { MESSAGES } from '@/constants';

// Services
import { productAPI } from '@/services/apis';

// Stores
import { TCartStore, useCartStore } from '@/stores';

// Types
import { ICartData, IProduct, TResponse } from '@/interface';

export type TUseCart = {
  handleAddProductToCart: (product: IProduct) => boolean;
  handleRemove: (productId: number) => TResponse;
  handleQuantity: (productId: number, quantity: number) => Promise<TResponse>;
  handleCheckout: () => Promise<void>;
};

export const useHandleCart = (): TUseCart => {
  //  Get handle add to cart from store
  const addNewProduct = useCartStore(
    (state: TCartStore): TCartStore['addNewProduct'] => state.addNewProduct,
  );

  //  Get handle update cart
  const updateCart = useCartStore(
    (state: TCartStore): TCartStore['updateCart'] => state.updateCart,
  );

  //  Handle check quantity
  const isValidQuantity = useCallback(
    (quantityInCart: number, quantityInStock: number): boolean => {
      return (
        quantityInCart <= quantityInStock &&
        !!quantityInCart &&
        !!quantityInStock
      );
    },
    [],
  );

  // Handle add product to cart
  const handleAddProductToCart = useCallback(
    (product: IProduct): boolean => {
      try {
        //  Get data from cart store
        const { data: carts } = useCartStore.getState();
        const { id, description, imageURL, name, price, quantity } = product;

        // Check Product exists from cart store
        const isExist: ICartData | undefined = carts.find(
          (item) => item.productId === id,
        );

        //  Check quantity in stock is valid
        let isValid: boolean = isValidQuantity(quantity, quantity);

        if ((!isExist || isExist) && !isValid) return false;

        if (!isExist) {
          const newRecord: ICartData = {
            productId: id,
            quantity: 1,
            description,
            imageURL,
            name,
            price,
          };

          addNewProduct(newRecord);

          return true;
        }

        //  Check quantity between in stock and cart
        isValid = isValidQuantity(isExist.quantity, quantity);

        if (!isValid) return false;

        isExist.quantity += 1;
        isExist.price = price * isExist.quantity;

        updateCart([...carts]);

        return true;
      } catch (error) {
        return false;
      }
    },
    [addNewProduct, isValidQuantity, updateCart],
  );

  //  Handle remove product from cart
  const handleRemove = useCallback(
    (productId: number): TResponse => {
      try {
        const carts: ICartData[] = useCartStore.getState().data;
        const newCarts: ICartData[] = carts.filter(
          (cart: ICartData): boolean => cart.productId !== productId,
        );

        updateCart(newCarts);

        return {
          isError: false,
          message: MESSAGES.REMOVE_FORM_CART_SUCCESS,
        };
      } catch (error) {
        return {
          isError: true,
          message: MESSAGES.FAIL_TO_FETCH,
        };
      }
    },
    [updateCart],
  );

  //  Handle change quantity from cart
  const handleQuantity = useCallback(
    async (productId: number, quantity: number): Promise<TResponse> => {
      try {
        // Get product from BE
        const product: IProduct | undefined =
          await productAPI.getById(productId);

        if (!product) {
          return {
            isError: true,
            message: MESSAGES.FAIL_TO_FETCH,
          };
        }

        const { price, quantity: quantityInStock } = product;
        // Get carts from localStore
        const carts: ICartData[] = useCartStore.getState().data;
        // Get cart matching with productID
        const cart: ICartData = carts.find(
          (cart: ICartData) => cart.productId === productId,
        ) as ICartData;

        const isValid: boolean = isValidQuantity(quantity, quantityInStock);

        if (!isValid)
          return {
            isError: true,
            message: MESSAGES.CHANGE_QUANTITY_ERROR,
          };

        // Update information
        cart.price = price * quantity;
        cart.quantity = quantity;

        updateCart([...carts]);

        return {
          isError: false,
        };
      } catch (error) {
        console.log(error);

        const message: string = (error as unknown as Error).message;

        return {
          isError: true,
          message: message,
        };
      }
    },
    [isValidQuantity, updateCart],
  );

  //  Handle checkout
  const handleCheckout = useCallback(async () => {
    // Get products from DB
    const products: IProduct[] = (await productAPI.getAll()) || [];
    //  Get cart from localStore
    const carts: ICartData[] = useCartStore.getState().data;

    await Promise.all(
      carts.map((cart: ICartData) => {
        const { quantity, productId } = cart;
        const product: IProduct | undefined = products.find(
          (product: IProduct) => product.id === productId,
        );

        if (!product) return Promise.resolve();

        return productAPI.update(productId, {
          quantity: product.quantity - quantity,
        });
      }),
    );
  }, []);

  return {
    handleAddProductToCart,
    handleRemove,
    handleQuantity,
    handleCheckout,
  };
};
