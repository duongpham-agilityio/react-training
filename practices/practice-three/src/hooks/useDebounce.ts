import { useCallback, useRef } from 'react';

// Constants
import { TIMES } from '@/constants';

export type TUseDebounce<T> = (value: T) => void;

/**
 * - This is a custom hook that listens for user input
 * @param callback Have a string as parameter
 * @returns
 */
export const useDebounce = <T>(callback?: TUseDebounce<T>): TUseDebounce<T> => {
  const refTime = useRef<ReturnType<typeof setTimeout>>();

  const debounce = useCallback(
    (value: T): void => {
      if (refTime.current) clearTimeout(refTime.current);

      refTime.current = setTimeout(() => {
        if (callback) callback(value);
      }, TIMES.DEBOUNCE);
    },
    [callback],
  );

  return debounce;
};
