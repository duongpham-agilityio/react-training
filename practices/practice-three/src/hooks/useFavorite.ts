import { create } from 'zustand';
import { createJSONStorage, persist } from 'zustand/middleware';

// Styles
import { IProduct } from '@/interface';

export interface IUseFavorite {
  data: IProduct[];
  onToggleFavorite: (product?: IProduct) => void;
}

export const useFavorite = create(
  persist<IUseFavorite>(
    (set, get) => ({
      data: [],
      onToggleFavorite: (product?: IProduct): void => {
        const { data } = get();

        if (!product) {
          return;
        }

        // check product exists in the store.
        const isExist: boolean = !!data.find((item) => product.id === item.id);

        // Remove from favorites list when Product already exists
        if (isExist) {
          const newFavorites = data.filter((item) => product.id !== item.id);

          return set({ data: newFavorites });
        }

        // Add product to favorites
        return set({
          data: [...data, product],
        });
      },
    }),
    {
      name: 'favorites',
      storage: createJSONStorage(() => localStorage),
    },
  ),
);
