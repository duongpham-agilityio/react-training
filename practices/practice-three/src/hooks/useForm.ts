import { FormEvent, useCallback, useState } from 'react';

type TFormData<T> = T & { [key: string]: string };
type TSubmitCallback = () => void;
type TUseForm<T> = {
  formData: TFormData<T>;
  onChange: (value: string, name: string) => void;
  onSubmit: (callback: TSubmitCallback) => void;
};

export const useForm = <T>(value: TFormData<T>): TUseForm<T> => {
  const [formData, setFormData] = useState<TFormData<T>>(value);

  const onChange = useCallback(
    (value: string, name: string) =>
      setFormData((prev) => ({ ...prev, [name]: value })),
    [],
  );

  const onSubmit = useCallback((callback: TSubmitCallback) => {
    return (e: FormEvent) => {
      e.preventDefault();

      callback();
    };
  }, []);

  return {
    formData,
    onChange,
    onSubmit,
  };
};
