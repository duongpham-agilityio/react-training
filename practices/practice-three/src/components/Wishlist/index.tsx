import { Box, Grid, GridItem } from '@chakra-ui/react';
import { memo, useCallback } from 'react';

// Hooks
import { useFavorite } from '@/hooks';

// Stores
import { IFavoriteStore, favoriteStore } from '@/stores';

// Constants
import { LIMIT_QUANTITY, MESSAGES } from '@/constants';

// Components
import { IProductCard, ProductCard } from '@/components';
import { Message } from '@/components/common';

// Types
import { IProduct } from '@/interface';

const Component = (): JSX.Element => {
  const favorites = favoriteStore((state: IFavoriteStore) => state.data);
  const { onToggleFavorite } = useFavorite();

  // Check for product is liked
  const isLiked = useCallback(
    (id: number): boolean => {
      return !!favorites.find((item) => id === item.id);
    },
    [favorites],
  );

  const handleAddToFavorite = useCallback(
    (id: number) => {
      const product: IProduct | undefined = favorites.find(
        (item) => item.id === id,
      );

      onToggleFavorite(product);
    },
    [favorites, onToggleFavorite],
  );

  // Todo: Update to late
  const handleAddToCart = useCallback(() => {}, []);

  const handleRenderProduct = useCallback(
    (product: IProduct): JSX.Element => {
      const { id, imageURL, name, description, price, quantity } = product;
      const info: IProductCard = {
        id,
        imageURL,
        price,
        description,
        title: name,
        status: quantity <= LIMIT_QUANTITY,
        statusMessage: `Only ${quantity} left`,
        isLiked: isLiked(id),
      };

      return (
        <GridItem key={id}>
          <ProductCard
            info={info}
            onLike={handleAddToFavorite}
            onAddToCart={handleAddToCart}
          />
        </GridItem>
      );
    },
    [handleAddToCart, handleAddToFavorite, isLiked],
  );

  return (
    <Box overflowY="scroll" h="full" py={5}>
      <Grid
        gridTemplateColumns={{
          base: '1fr',
          '2xl': '1fr 1fr',
        }}
        gap={5}
      >
        {favorites.length ? (
          favorites.map(handleRenderProduct)
        ) : (
          <Message message={MESSAGES.EMPTY} />
        )}
      </Grid>
    </Box>
  );
};

const Wishlist = memo(Component);

export default Wishlist;
