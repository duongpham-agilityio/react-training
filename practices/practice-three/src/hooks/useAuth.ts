import { useCallback, useState } from 'react';
import { useNavigate } from 'react-router-dom';

// Hooks
import { useToast } from '@/hooks';

// Stores
import { TAuthStore, useAuthStore } from '@/stores';

// Constants
import { MESSAGES, ROUTES, TITLES } from '@/constants';

export type TUseHandleAuth<T> = {
  isError: boolean;
  error: T;
  onLogin: (email: string, password: string) => void;
  onLogout: () => void;
};

type TUseAuthError<T> = {
  isError: boolean;
  error: T;
};

export const useHandleAuth = <T extends object>(): TUseHandleAuth<T> => {
  // Redirect to other page
  const redirect = useNavigate();

  // Show toast
  const { showToast } = useToast();

  // Error
  const [err, setError] = useState<TUseAuthError<T>>({
    isError: false,
    error: {} as T,
  });

  const handleSetAuth = useAuthStore((state: TAuthStore) => state.setIsAuth);
  const handleClearAuth = useAuthStore(
    (state: TAuthStore) => state.clearIsAuth,
  );

  const onLogin: TUseHandleAuth<T>['onLogin'] = useCallback(
    async (email: string, password: string) => {
      try {
        const response: Response = await fetch('', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ email, password }),
        });

        if (response.status === 404) throw new Error(MESSAGES.FAIL_TO_FETCH);

        if (response.status === 401) {
          const data = await response.json();

          return setError({
            isError: true,
            error: data.data,
          });
        }

        handleSetAuth();
        redirect(ROUTES.ROOT);
      } catch (error) {
        const message: string = (error as unknown as Error).message;

        showToast({
          title: TITLES.ERROR,
          description: message,
          status: 'error',
        });
      }
    },
    [handleSetAuth, redirect, showToast],
  );

  const onLogout = useCallback(() => handleClearAuth(), [handleClearAuth]);

  return { ...err, onLogin, onLogout };
};
