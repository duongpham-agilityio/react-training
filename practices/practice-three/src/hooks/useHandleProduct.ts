import { useCallback } from 'react';
import { useQueryClient } from '@tanstack/react-query';

// Types
import { IFormAddData } from '@/pages/Profile/components/FormAdd';

// Services
import { ProductPayload, productAPI } from '@/services/apis';

// Constants
import { ENDPOINT_SERVICES } from '@/constants';

export type TUseHandleProduct = {
  onAddProduct: (data: IFormAddData) => Promise<boolean>;
  onUpdateProduct: (
    id: number,
    product: Partial<ProductPayload>,
  ) => Promise<boolean>;
  onRemoveProduct: (id: number) => Promise<boolean>;
};

export const useHandleProduct = (): TUseHandleProduct => {
  const queryClient = useQueryClient();

  const onAddProduct = useCallback(
    async (data: IFormAddData): Promise<boolean> => {
      try {
        const product: ProductPayload = {
          ...data,
          isLiked: false,
        };
        await productAPI.add(product);

        queryClient.invalidateQueries([ENDPOINT_SERVICES.PRODUCTS]);

        return true;
      } catch (error) {
        return false;
      }
    },
    [queryClient],
  );

  const onUpdateProduct = useCallback(
    async (id: number, product: Partial<ProductPayload>): Promise<boolean> => {
      try {
        await productAPI.update(id, product);
        queryClient.invalidateQueries([ENDPOINT_SERVICES.PRODUCTS]);

        return true;
      } catch (error) {
        return false;
      }
    },
    [queryClient],
  );

  const onRemoveProduct = useCallback(
    async (id: number): Promise<boolean> => {
      try {
        await productAPI.removeById(id);
        queryClient.invalidateQueries([ENDPOINT_SERVICES.PRODUCTS]);

        return true;
      } catch (error) {
        return false;
      }
    },
    [queryClient],
  );

  return {
    onAddProduct,
    onUpdateProduct,
    onRemoveProduct,
  };
};
