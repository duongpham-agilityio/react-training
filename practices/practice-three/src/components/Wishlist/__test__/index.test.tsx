import { fireEvent, render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Hooks
import { useFavorite } from '@/hooks';

// Components
import Wishlist from '..';

const setup = () =>
  render(
    <BrowserRouter>
      <Wishlist />
    </BrowserRouter>,
  );

describe('Wishlist', () => {
  beforeEach(() => {
    useFavorite.setState({
      data: [
        {
          name: 'Tasty Plastic Mouse',
          description: 'Quos tempora adipisci quidem ipsum.',
          imageURL: 'https://loremflickr.com/640/480/fashion',
          price: 430,
          quantity: 10,
          isLiked: true,
          id: 1,
          category: 'fashion',
          createdAt: new Date(),
        },
        {
          name: 'Tasty Plastic Mouse',
          description: 'Quos tempora adipisci quidem ipsum.',
          imageURL: 'https://loremflickr.com/640/480/fashion',
          price: 430,
          quantity: 10,
          isLiked: true,
          id: 2,
          category: 'books',
          createdAt: new Date(),
        },
      ],
    });
  });

  it('Match to snapshot', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('Remove from wishlist', () => {
    const { container } = setup();
    const buttons = container.querySelectorAll(
      'button[ aria-label="Button add to favorite"]',
    );

    expect(useFavorite.getState().data.length).toBe(2);

    fireEvent.click(buttons[0]);

    expect(useFavorite.getState().data.length).toBe(1);
  });
});
