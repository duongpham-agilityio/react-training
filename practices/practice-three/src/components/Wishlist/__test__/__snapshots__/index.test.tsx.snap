// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wishlist Match to snapshot 1`] = `
<div>
  <div
    class="css-1jk9vzu"
  >
    <div
      class="css-1d6fzre"
    />
  </div>
</div>
`;
